cupyx/scipy/sparse/_index.py

```diff
@@ -23,6 +23,155 @@
 _bool_scalar_types = (bool, numpy.bool, numpy.bool_)
 
 
+def _csr_column_index1_indptr(unique_idxs, sort_idxs, col_counts,
+                              Ap, Aj):
+    """Construct output indptr by counting column indices
+    in input matrix for each row.
+    Args
+        unique_idxs : Unique set of indices sorted in ascending order
+        sort_idxs : Indices sorted to preserve original order of unique_idxs
+        col_counts : Number of times each unique index occurs in Aj
+        Ap : indptr array of input sparse matrix
+        Aj : indices array of input sparse matrix
+    Returns
+        Bp : Output indptr
+        Aj_mask : Input indices array with all cols not matching the
+                  index masked out with -1.
+    """
+    out_col_sum = cupy.zeros((Aj.size+1,), dtype=col_counts.dtype)
+
+    index = cupy.argsort(unique_idxs)
+    sorted_index = cupy.searchsorted(unique_idxs, Aj)
+
+    yindex = cupy.take(index, sorted_index)
+    mask = unique_idxs[yindex] == Aj
+
+    idxs_adj = _csr_column_inv_idx(unique_idxs)
+    out_col_sum[1:][mask] = col_counts[idxs_adj[Aj[mask]]]
+
+    Aj_mask = out_col_sum[1:].copy()
+    Aj_mask[Aj_mask == 0] = -1
+
+    Aj_mask[Aj_mask > 0] = Aj[Aj_mask > 0]
+    Aj_mask[Aj_mask > 0] = cupy.searchsorted(
+        unique_idxs, Aj_mask[Aj_mask > 0])
+
+    Aj_mask[Aj_mask >= 0] = sort_idxs[Aj_mask[Aj_mask >= 0]]
+
+    cupy.cumsum(out_col_sum, out=out_col_sum)
+    Bp = out_col_sum[Ap]
+    Bp[1:] -= Bp[:-1]
+    cupy.cumsum(Bp, out=Bp)
+
+    return Bp, Aj_mask
+
+
+def _csr_column_index1(col_idxs, Ap, Aj):
+    """Construct indptr and components for populating indices and data of
+    output sparse array
+    Args
+        col_idxs : column indices to index from input indices
+        Ap : indptr of input sparse matrix
+        Aj : indices of input sparse matrix
+    Returns
+        Bp : indptr of output sparse matrix
+        Aj_mask : Input indices array with all cols not matching the index
+                  index masked out with -1.
+        col_counts : Number of times each unique index occurs in Aj
+        sort_idxs : Indices sorted to preserve original order of idxs
+    """
+
+    idx_map, sort_idxs = cupy.unique(col_idxs, return_index=True)
+    sort_idxs = sort_idxs.astype(idx_map.dtype)
+    idxs = cupy.searchsorted(idx_map, col_idxs)
+
+    col_counts = cupy.zeros(idx_map.size, dtype=col_idxs.dtype)
+    cupyx.scatter_add(col_counts, idxs, 1)
+
+    Bp, Aj_mask = _csr_column_index1_indptr(
+        idx_map, sort_idxs, col_counts, Ap, Aj)
+
+    return Bp, Aj_mask, col_counts, sort_idxs
+
+
+_csr_column_index2_ker = core.ElementwiseKernel(
```

Comment:
This kernel looks to work slowly in some cases for the same reasons as https://github.com/cupy/cupy/pull/3658#discussion_r470135917, but it's OK to leave in this PR.